Fix Chrome window.innerWidth priority for correct canvas resizing

ayushman1210 opened the pull request:
Resolves #8121 

Changes:
- Modified the `getWindowWidth()` and `getWindowHeight()` functions in `environment.js` to prioritize `document.documentElement.clientWidth` and `document.documentElement.clientHeight` over `window.innerWidth` and `window.innerHeight`.
- This resolves an issue in Chrome on Android where `windowWidth` and `windowHeight` values do not update correctly when the device is rotated between portrait and landscape orientation.
- Updated logic ensures consistent behavior across mobile browsers while maintaining backward compatibility with desktop browsers.

#### PR Checklist

- [x] `npm run lint` passes
- [x] Inline reference is included / updated
- [x] Unit tests are included / updated

[Inline reference]: https://p5js.org/contribute/contributing_to_the_p5js_reference/
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

ayushman1210 committed:
Fix Chrome window.innerWidth priority for correct canvas resizing

ayushman1210 committed:
fixed the code

eyaler commented:
@ayushman1210 thanks! I think innerWidth should be before body.clientWidth as the body dimension reflects the content and not the viewport. Also, the suggest change may have issue when scrollbars are present (in Chrome they are part of the innerWidth but not of the documentElement.clientWidth) - I did not look into this yet.

perminder-17 reviewed (changes requested):
Can you explain what's the reason behind removing all the docs? I think you need to fix the PR with the minimal changes whatever needed.

eyaler commented:
@ayushman1210 i don't get it. this was the state of things before with the issue that innerWidth may be wrong, no? my suggestion was that document.documentElement.clientWidth will be used first (but i think we need more feedback before making this change)

ayushman1210 committed:
fixed the code

ayushman1210 committed:
fixed the code added visual-viewport

eyaler commented:
thanks @ayushman1210 

>     * I fixed the timing issue by deferring the resize handler to the next animation frame so layout has settled after rotation (so windowResized() sees the post-rotation viewport).

I coundn't find the added requestAnimationFrame in your PR, and in my own testing requestAnimationFrame did not solve the issue. Can you point me to the commit/line?

>     * I also updated getWindowWidth() / getWindowHeight() to prefer visualViewport (when available) and then fall back to other viewport measures

In my testing visualViewport does not include scrollbars as documentElement.clientWidth, and also is wrong when rotating back just is innerWidth. I did find viewport.segments[0].width that seems to both include scrollbars and be correct after orientation change but that is still experimental and has partial support (and i don't have a good enough understanding of it)

>     * I agree that preferring [document.documentElement.clientWidth] avoids surprising inclusion of scrollbar width in many desktop Chrome cases — that was exactly your point.

On the contrary, I actully think scrollbar should be included

ayushman1210 committed:
env: prefer visualViewport.segments and include scrollbars when available

Prefer experimental visualViewport.segments[0] and visualViewport before window.inner* so measurements include scrollbars and handle orientation changes when available. Guard experimental access and fall back to documentElement.client* for compatibility.

ayushman1210 committed:
fixed the get width and getheight section

ayushman1210 commented:
@eyaler
You're absolutely right about the scrollbar issue. I've removed `visualViewport.width` from the fallback chain to avoid the inconsistency.

**Updated implementation:**
- Priority 1: `visualViewport.segments[0]` (experimental foldable/dual-screen API)
- Priority 2: `window.innerWidth` (includes scrollbars, matches historical p5.js behavior)
- Priority 3: `document.documentElement.clientWidth` (fallback when innerWidth unavailable)
- Priority 4: `document.body.clientWidth` (last resort)

The segments API is only used when explicitly available (for foldable devices), then we fall back directly to `innerWidth`. This avoids the Chrome scrollbar issue you mentioned while still supporting dual-screen devices.

Regarding `requestAnimationFrame`: I apologize for any confusion - I did NOT add or modify any RAF code. My changes are limited to the `getWindowWidth()` and `getWindowHeight()` helper functions, which are only called:
1. On initialization in `_updateWindowSize()` 
2. In the `_onresize` event handler

These functions are NOT called in the animation loop, so there's no per-frame performance impact.

Thanks 
Ayushman

ayushman1210 commented:
Hi @perminder-17 
can you please Review this PR #8122  for the issue #8121

ayushman1210 committed:
Merge branch 'main' into fix/chrome-window-innerwidth

ayushman1210 committed:
Merge branch 'main' into fix/chrome-window-innerwidth

perminder-17 reviewed (approved):
Sorry for the delay in review, I'll be testing the changes once.

perminder-17 committed:
running CI

perminder-17 commented on src/core/environment.js:
```diff
@@ -782,7 +782,39 @@ p5.prototype._onresize = function(e) {
   }
 };
 
+// Prefer the visual viewport when available (better for mobile: pinch/keyboard/zoom),
+// then the layout viewport, then the window inner size, then the body content size.
+// Order chosen intentionally:
+// 1) window.visualViewport.*         -> visual viewport (mobile-friendly; adapts to on-screen keyboard)
+// 2) document.documentElement.client* -> layout viewport (avoids including scrollbars on some platforms)
+// 3) window.inner*                   -> includes scrollbars on some platforms (Chrome)
+// 4) document.body.client*           -> content-box fallback if the above are unavailable
+// Caveat: different browsers expose subtly different viewport measurements (scrollbars, zoom,
+// and visual viewport). This ordering prefers mobile-friendly measurements while retaining
+// sensible fallbacks for desktop browsers.
 function getWindowWidth() {
+  // Prefer including scrollbars when possible per contributor guidance.
+  // Order: experimental visualViewport.segments (may include scrollbars/handle rotation),
+  // visualViewport.width, window.innerWidth (includes scrollbars),
+  // documentElement.clientWidth (excludes scrollbars), document.body.clientWidth.
+  try {
+    if (window.visualViewport) {
+      const segs = window.visualViewport.segments;
+      if (Array.isArray(segs) && segs.length > 0) {
+        const w = segs[0] && typeof segs[0].width === 'number' ? segs[0].width : null;
+        if (w && w > 0) {
+          return w;
+        }
+      }
+
+      // if (typeof window.visualViewport.width === 'number' && window.visualViewport.width > 0) {
+      //   return window.visualViewport.width;
+      // }
+    }
+  } catch (e) {
+    // experimental access may throw; ignore and continue with fallbacks
+  }
+
   return (
     window.innerWidth ||
```
The extra `visualViewport.segments` part and the `try/catch` around it aren't really needed for this, which isn't what this bug is about. There are also a few commented-out lines left in `getWindowWidth/getWindowHeight` that we can remove to keep things tidy.

So it could come down to just this one line:

```js
(document.documentElement && document.documentElement.clientWidth) || window.innerWidth || (document.body && document.body.clientWidth) || 0
```

perminder-17 reviewed (changes requested):
Hi @ayushman1210 @eyaler , really sorry for the delay in reviewing this one. 

Btw, are you still able to reproduce this issue with the latest chrome. Looks like I am unable to reproduce this issue on my android chrome. 

If the issue is still reproduceable, @eyaler gave a very nice finding;

> the issue seems to be with chrome window.innerWidth, so maybe in https://github.com/processing/p5.js/blob/main/src/core/environment.js#L787 the order should be changes so that innerWidth gets lower priority?

The main thing the issue needs is just changing the order in which we check the window size,  checking `document.documentElement.clientWidth` first, before `window.innerWidth`, since `innerWidth` is the one that gives the wrong value on Chrome Android after rotating.

perminder-17 reviewed (commented):
For now, I am closing this PR since I am unable to reproduce the issue. Will be reopening if we again face the same issue, but looks like the latest chrome doesn't have that issue.
